engine/crates/runtime/src/kv.rs

```diff
@@ -1,37 +1,47 @@
-use std::time::Duration;
+use std::{sync::Arc, time::Duration};
 
 use serde::{de::DeserializeOwned, Serialize};
 
+#[derive(Debug, thiserror::Error)]
+pub enum KvError {
+    #[error("Kv error: {0}")]
+    Kv(String),
+    #[error("Serialization error: {0}")]
+    Serialization(#[from] serde_json::Error),
+}
+
+pub type KvResult<T> = Result<T, KvError>;
+
+// used in gateway-core for jwt-verifier, not sure if relevant outside.
 pub trait KvManager {
-    type Error: std::fmt::Debug;
-    type Kv: KvStore<Error = Self::Error>;
+    fn load<V: Serialize + DeserializeOwned + Send + Sync + 'static>(&self, namespace: &str) -> KvResult<KvStore<V>>;
+}
 
-    fn load(&self, namespace: &str) -> Result<Self::Kv, Self::Error>;
+/// Not used during the request execution but while building the engine to avoid
+/// exposing the Value type.
+pub trait KvStoreFactory: Send + Sync {
+    fn create<V: Clone + Serialize + DeserializeOwned + Send + Sync + 'static>(&self) -> KvResult<KvStore<V>>;
```

Comment:
I've made comments elsewhere/discussed in slack, but would be good to get rid of the factory if poss?

Can have an inner dyn safe trait that just returns a `Vec<u8>` then build generic stuff on top of that to do the decoding, or similar.  Think it would clean up a bunch